--- a/drivers/net/wireless/ath/ath9k/ar953x_initvals.h	2015-01-26 10:48:30.536035102 +0800
+++ b/drivers/net/wireless/ath/ath9k/ar953x_initvals.h	2015-01-26 09:40:19.784163182 +0800
@@ -260,7 +260,7 @@
 	{0x0000a398, 0x001f0e0f},
 	{0x0000a39c, 0x0075393f},
 	{0x0000a3a0, 0xb79f6427},
-	{0x0000a3a4, 0x000000ff},
+	{0x0000a3a4, 0x00000000},//vicharl modify 0x000000ff ->0x00000000
 	{0x0000a3a8, 0x6a6a6a6a},
 	{0x0000a3ac, 0x6a6a6a6a},
 	{0x0000a3b0, 0x00c8641a},
@@ -287,11 +287,21 @@
 	{0x0000a404, 0x00000000},
 	{0x0000a408, 0x0e79e5c6},
 	{0x0000a40c, 0x00820820},
+	/* BEGIN: Modified by vicharl,openwrt 2014/8/28 */
+#if 0
 	{0x0000a414, 0x1ce42108},
 	{0x0000a418, 0x2d001dce},
 	{0x0000a41c, 0x1ce73908},
 	{0x0000a420, 0x000001ce},
 	{0x0000a424, 0x1ce738e7},
+#else
+	{0x0000a414, 0x1ce739ce}, // BB_tpc_8 
+	{0x0000a418, 0x2d001dce},
+	{0x0000a41c, 0x1ce739ce}, // BB_tpc_10 
+	{0x0000a420, 0x000001ce},
+	{0x0000a424, 0x1ce739ce   }, // BB_tpc_12 
+#endif
+	/* END:   Modified by vicharl,openwrt 2014/8/28 */
 	{0x0000a428, 0x000001ce},
 	{0x0000a42c, 0x1ce739ce},
 	{0x0000a430, 0x1ce739ce},
@@ -304,6 +314,8 @@
 	{0x0000a450, 0x00010000},
 	{0x0000a458, 0x00000000},
 	{0x0000a644, 0xbfad9d74},
+/* BEGIN: Modified by vicharl,openwrt 2014/8/28 */
+#if 0		
 	{0x0000a648, 0x0048060a},
 	{0x0000a64c, 0x00003c37},
 	{0x0000a670, 0x03020100},
@@ -313,6 +325,18 @@
 	{0x0000a680, 0x31301514},
 	{0x0000a684, 0x35343332},
 	{0x0000a688, 0x00000036},
+#else
+	 { 0x0000a648 ,  0x0048660a   }, // BB_txiqcal_control_1 
+ 	 { 0x0000a64c ,  0x00003c37   }, // BB_txiqcal_control_2 
+ 	 { 0x0000a670 ,  0x03020100   }, // BB_cal_rxbb_gain_tbl_0 
+  	{ 0x0000a674 ,  0x21200504   }, // BB_cal_rxbb_gain_tbl_4 
+  	{ 0x0000a678 ,  0x61602322   }, // BB_cal_rxbb_gain_tbl_8 
+  	{ 0x0000a67c ,  0x65646362   }, // BB_cal_rxbb_gain_tbl_12 
+  	{ 0x0000a680 ,  0x6B6A6968   }, // BB_cal_rxbb_gain_tbl_16 
+ 	 { 0x0000a684 ,  0xE2706D6C   }, // BB_cal_rxbb_gain_tbl_20 
+  	{ 0x0000a688 ,  0x000000e3   }, // BB_cal_rxbb_gain_tbl_24 
+#endif
+/* END:   Modified by vicharl,openwrt 2014/8/28 */
 	{0x0000a690, 0x08000838},
 	{0x0000a7cc, 0x00000000},
 	{0x0000a7d0, 0x00000000},
@@ -715,4 +739,207 @@
 	{0x00016448, 0x6c927a70},
 };
 
+
+/* BEGIN: Added by vicharl,openwrt  2014/8/28 */
+static const u_int32_t qca953xModes_w_xpa_tx_gain_table_honeybee_1p1[][2] = {
+/*    Addr        allmodes        */
+  { 0x0000a2dc ,  0xcc69ac94   }, // BB_cl_map_0_b0 
+  { 0x0000a2e0 ,  0xf0b23118   }, // BB_cl_map_1_b0 
+  { 0x0000a2e4 ,  0xffffc000   }, // BB_cl_map_2_b0 
+  { 0x0000a2e8 ,  0xc0000000   }, // BB_cl_map_3_b0 
+  { 0x0000a410 ,  0x000050d2   }, // BB_tpc_7 
+  { 0x0000a500 ,  0x09000040   }, // BB_tx_gain_tab_1 
+  { 0x0000a504 ,  0x0b000041   }, // BB_tx_gain_tab_2 
+  { 0x0000a508 ,  0x0d000042   }, // BB_tx_gain_tab_3 
+  { 0x0000a50c ,  0x11000044   }, // BB_tx_gain_tab_4 
+  { 0x0000a510 ,  0x15000046   }, // BB_tx_gain_tab_5 
+  { 0x0000a514 ,  0x1d000440   }, // BB_tx_gain_tab_6 
+  { 0x0000a518 ,  0x1f000441   }, // BB_tx_gain_tab_7 
+  { 0x0000a51c ,  0x23000443   }, // BB_tx_gain_tab_8 
+  { 0x0000a520 ,  0x25000444   }, // BB_tx_gain_tab_9 
+  { 0x0000a524 ,  0x28000a40   }, // BB_tx_gain_tab_10 
+  { 0x0000a528 ,  0x2c000a42   }, // BB_tx_gain_tab_11 
+  { 0x0000a52c ,  0x2e000a43   }, // BB_tx_gain_tab_12 
+  { 0x0000a530 ,  0x30000a44   }, // BB_tx_gain_tab_13 
+  { 0x0000a534 ,  0x34000a46   }, // BB_tx_gain_tab_14 
+  { 0x0000a538 ,  0x37000ce0   }, // BB_tx_gain_tab_15 
+  { 0x0000a53c ,  0x3b000ce2   }, // BB_tx_gain_tab_16 
+  { 0x0000a540 ,  0x3d000ce3   }, // BB_tx_gain_tab_17 
+  { 0x0000a544 ,  0x3f000ce4   }, // BB_tx_gain_tab_18 
+  { 0x0000a548 ,  0x45001ee0   }, // BB_tx_gain_tab_19 
+  { 0x0000a54c ,  0x49001ee2   }, // BB_tx_gain_tab_20 
+  { 0x0000a550 ,  0x4d001ee4   }, // BB_tx_gain_tab_21 
+  { 0x0000a554 ,  0x51001ee6   }, // BB_tx_gain_tab_22 
+  { 0x0000a558 ,  0x55001eea   }, // BB_tx_gain_tab_23 
+  { 0x0000a55c ,  0x59001eec   }, // BB_tx_gain_tab_24 
+  { 0x0000a560 ,  0x5d001ef0   }, // BB_tx_gain_tab_25 
+  { 0x0000a564 ,  0x5f001ef1   }, // BB_tx_gain_tab_26 
+  { 0x0000a568 ,  0x60001ef2   }, // BB_tx_gain_tab_27 
+  { 0x0000a56c ,  0x61001ef3   }, // BB_tx_gain_tab_28 
+  { 0x0000a570 ,  0x62001ef4   }, // BB_tx_gain_tab_29 
+  { 0x0000a574 ,  0x63001ef5   }, // BB_tx_gain_tab_30 
+  { 0x0000a578 ,  0x64001ffc   }, // BB_tx_gain_tab_31 
+  { 0x0000a57c ,  0x64001ffc   }, // BB_tx_gain_tab_32 
+  { 0x0000a600 ,  0x00000000   }, // BB_caltx_gain_set_0 
+  { 0x0000a604 ,  0x00000000   }, // BB_caltx_gain_set_2 
+  { 0x0000a608 ,  0x00000000   }, // BB_caltx_gain_set_4 
+  { 0x0000a60c ,  0x00000000   }, // BB_caltx_gain_set_6 
+  { 0x0000a610 ,  0x00000000   }, // BB_caltx_gain_set_8 
+  { 0x0000a614 ,  0x00000000   }, // BB_caltx_gain_set_10 
+  { 0x0000a618 ,  0x00000000   }, // BB_caltx_gain_set_12 
+  { 0x0000a61c ,  0x00000201   }, // BB_caltx_gain_set_14 
+  { 0x0000a620 ,  0x00000201   }, // BB_caltx_gain_set_16 
+  { 0x0000a624 ,  0x00000201   }, // BB_caltx_gain_set_18 
+  { 0x0000a628 ,  0x00000201   }, // BB_caltx_gain_set_20 
+  { 0x0000a62c ,  0x00000201   }, // BB_caltx_gain_set_22 
+  { 0x0000a630 ,  0x00000201   }, // BB_caltx_gain_set_24 
+  { 0x0000a634 ,  0x00000201   }, // BB_caltx_gain_set_26 
+  { 0x0000a638 ,  0x00000201   }, // BB_caltx_gain_set_28 
+  { 0x0000a63c ,  0x00000e02   }, // BB_caltx_gain_set_30 
+  { 0x0000b2dc ,  0xcc69ac94   }, // BB_cl_map_0_b1 
+  { 0x0000b2e0 ,  0xf0b23118   }, // BB_cl_map_1_b1 
+  { 0x0000b2e4 ,  0xffffc000   }, // BB_cl_map_2_b1 
+  { 0x0000b2e8 ,  0xc0000000   }, // BB_cl_map_3_b1 
+  { 0x00016044 ,  0x054922DB   }, // ch0_TXRF2 
+  { 0x00016048 ,  0x6c927a70   }, // ch0_TXRF3 
+  { 0x00016444 ,  0x054922DB   }, // ch1_TXRF2 
+  { 0x00016448 ,  0x6c927a70   }, // ch1_TXRF3 
+  { 0x00016284 ,  0x15D3F000   }, // ch0_TOP2
+};
+
+
+static const u_int32_t qca953xModes_w_xpa_tx_gain_table_honeybee_2p0[][2] = {
+/*    Addr        allmodes        */
+  { 0x0000a2dc ,  0xfffb52aa   }, // BB_cl_map_0_b0 
+  { 0x0000a2e0 ,  0xfffd64cc   }, // BB_cl_map_1_b0 
+  { 0x0000a2e4 ,  0xfffe80f0   }, // BB_cl_map_2_b0 
+  { 0x0000a2e8 ,  0xffffff00   }, // BB_cl_map_3_b0 
+  { 0x0000a410 ,  0x000050d5   }, // BB_tpc_7 
+  { 0x0000a500 ,  0x00000000   }, // BB_tx_gain_tab_1 
+  { 0x0000a504 ,  0x04000002   }, // BB_tx_gain_tab_2 
+  { 0x0000a508 ,  0x08000004   }, // BB_tx_gain_tab_3 
+  { 0x0000a50c ,  0x0c000006   }, // BB_tx_gain_tab_4 
+  { 0x0000a510 ,  0x1000000a   }, // BB_tx_gain_tab_5 
+  { 0x0000a514 ,  0x1400000c   }, // BB_tx_gain_tab_6 
+  { 0x0000a518 ,  0x1800000e   }, // BB_tx_gain_tab_7 
+  { 0x0000a51c ,  0x1c000048   }, // BB_tx_gain_tab_8 
+  { 0x0000a520 ,  0x2000004a   }, // BB_tx_gain_tab_9 
+  { 0x0000a524 ,  0x2400004c   }, // BB_tx_gain_tab_10 
+  { 0x0000a528 ,  0x2800004e   }, // BB_tx_gain_tab_11 
+  { 0x0000a52c ,  0x2b00024a   }, // BB_tx_gain_tab_12 
+  { 0x0000a530 ,  0x2f00024c   }, // BB_tx_gain_tab_13 
+  { 0x0000a534 ,  0x3300024e   }, // BB_tx_gain_tab_14 
+  { 0x0000a538 ,  0x36000668   }, // BB_tx_gain_tab_15 
+  { 0x0000a53c ,  0x38000669   }, // BB_tx_gain_tab_16 
+  { 0x0000a540 ,  0x3a000868   }, // BB_tx_gain_tab_17 
+  { 0x0000a544 ,  0x3d00086a   }, // BB_tx_gain_tab_18 
+  { 0x0000a548 ,  0x4000086c   }, // BB_tx_gain_tab_19 
+  { 0x0000a54c ,  0x4200086e   }, // BB_tx_gain_tab_20 
+  { 0x0000a550 ,  0x43000a6e   }, // BB_tx_gain_tab_21 
+  { 0x0000a554 ,  0x43000a6e   }, // BB_tx_gain_tab_22 
+  { 0x0000a558 ,  0x43000a6e   }, // BB_tx_gain_tab_23 
+  { 0x0000a55c ,  0x43000a6e   }, // BB_tx_gain_tab_24 
+  { 0x0000a560 ,  0x43000a6e   }, // BB_tx_gain_tab_25 
+  { 0x0000a564 ,  0x43000a6e   }, // BB_tx_gain_tab_26 
+  { 0x0000a568 ,  0x43000a6e   }, // BB_tx_gain_tab_27 
+  { 0x0000a56c ,  0x43000a6e   }, // BB_tx_gain_tab_28 
+  { 0x0000a570 ,  0x43000a6e   }, // BB_tx_gain_tab_29 
+  { 0x0000a574 ,  0x43000a6e   }, // BB_tx_gain_tab_30 
+  { 0x0000a578 ,  0x43000a6e   }, // BB_tx_gain_tab_31 
+  { 0x0000a57c ,  0x43000a6e   }, // BB_tx_gain_tab_32 
+  { 0x0000a600 ,  0x00000000   }, // BB_caltx_gain_set_0 
+  { 0x0000a604 ,  0x00000000   }, // BB_caltx_gain_set_2 
+  { 0x0000a608 ,  0x00000000   }, // BB_caltx_gain_set_4 
+  { 0x0000a60c ,  0x03804000   }, // BB_caltx_gain_set_6 
+  { 0x0000a610 ,  0x03804e01   }, // BB_caltx_gain_set_8 
+  { 0x0000a614 ,  0x03804e01   }, // BB_caltx_gain_set_10 
+  { 0x0000a618 ,  0x03804e01   }, // BB_caltx_gain_set_12 
+  { 0x0000a61c ,  0x04009002   }, // BB_caltx_gain_set_14 
+  { 0x0000a620 ,  0x04009002   }, // BB_caltx_gain_set_16 
+  { 0x0000a624 ,  0x04009002   }, // BB_caltx_gain_set_18 
+  { 0x0000a628 ,  0x04009002   }, // BB_caltx_gain_set_20 
+  { 0x0000a62c ,  0x04009002   }, // BB_caltx_gain_set_22 
+  { 0x0000a630 ,  0x04009002   }, // BB_caltx_gain_set_24 
+  { 0x0000a634 ,  0x04009002   }, // BB_caltx_gain_set_26 
+  { 0x0000a638 ,  0x04009002   }, // BB_caltx_gain_set_28 
+  { 0x0000a63c ,  0x04009002   }, // BB_caltx_gain_set_30 
+  { 0x0000b2dc ,  0xfffb52aa   }, // BB_cl_map_0_b1 
+  { 0x0000b2e0 ,  0xfffd64cc   }, // BB_cl_map_1_b1 
+  { 0x0000b2e4 ,  0xfffe80f0   }, // BB_cl_map_2_b1 
+  { 0x0000b2e8 ,  0xffffff00   }, // BB_cl_map_3_b1 
+  { 0x00016044 ,  0x024922DB   }, // ch0_TXRF2 
+  { 0x00016048 ,  0x6c927a70   }, // ch0_TXRF3 
+  { 0x00016444 ,  0x024922DB   }, // ch1_TXRF2 
+  { 0x00016448 ,  0x6c927a70   }, // ch1_TXRF3 
+};
+
+
+
+static const u_int32_t qca953xModes_xpa_tx_gain_table_honeybee_1p0[][2] = {
+/*    Addr        allmodes        */
+  { 0x0000a2dc ,  0xfffd5aaa   }, // BB_cl_map_0_b0 
+  { 0x0000a2e0 ,  0xfffe9ccc   }, // BB_cl_map_1_b0 
+  { 0x0000a2e4 ,  0xffffe0f0   }, // BB_cl_map_2_b0 
+  { 0x0000a2e8 ,  0xfffcff00   }, // BB_cl_map_3_b0 
+  { 0x0000a410 ,  0x000050da   }, // BB_tpc_7 
+  { 0x0000a500 ,  0x00000000   }, // BB_tx_gain_tab_1 
+  { 0x0000a504 ,  0x04000002   }, // BB_tx_gain_tab_2 
+  { 0x0000a508 ,  0x08000004   }, // BB_tx_gain_tab_3 
+  { 0x0000a50c ,  0x0c000006   }, // BB_tx_gain_tab_4 
+  { 0x0000a510 ,  0x0f00000a   }, // BB_tx_gain_tab_5 
+  { 0x0000a514 ,  0x1300000c   }, // BB_tx_gain_tab_6 
+  { 0x0000a518 ,  0x1700000e   }, // BB_tx_gain_tab_7 
+  { 0x0000a51c ,  0x1b000064   }, // BB_tx_gain_tab_8 
+  { 0x0000a520 ,  0x1f000242   }, // BB_tx_gain_tab_9 
+  { 0x0000a524 ,  0x23000229   }, // BB_tx_gain_tab_10 
+  { 0x0000a528 ,  0x270002a2   }, // BB_tx_gain_tab_11 
+  { 0x0000a52c ,  0x2c001203   }, // BB_tx_gain_tab_12 
+  { 0x0000a530 ,  0x30001803   }, // BB_tx_gain_tab_13 
+  { 0x0000a534 ,  0x33000881   }, // BB_tx_gain_tab_14 
+  { 0x0000a538 ,  0x38001809   }, // BB_tx_gain_tab_15 
+  { 0x0000a53c ,  0x3a000814   }, // BB_tx_gain_tab_16 
+  { 0x0000a540 ,  0x3f001a0c   }, // BB_tx_gain_tab_17 
+  { 0x0000a544 ,  0x43001a0e   }, // BB_tx_gain_tab_18 
+  { 0x0000a548 ,  0x46001812   }, // BB_tx_gain_tab_19 
+  { 0x0000a54c ,  0x49001884   }, // BB_tx_gain_tab_20 
+  { 0x0000a550 ,  0x4d001e84   }, // BB_tx_gain_tab_21 
+  { 0x0000a554 ,  0x50001e69   }, // BB_tx_gain_tab_22 
+  { 0x0000a558 ,  0x550006f4   }, // BB_tx_gain_tab_23 
+  { 0x0000a55c ,  0x59000ad3   }, // BB_tx_gain_tab_24 
+  { 0x0000a560 ,  0x5e000ad5   }, // BB_tx_gain_tab_25 
+  { 0x0000a564 ,  0x61001ced   }, // BB_tx_gain_tab_26 
+  { 0x0000a568 ,  0x660018d4   }, // BB_tx_gain_tab_27 
+  { 0x0000a56c ,  0x660018d4   }, // BB_tx_gain_tab_28 
+  { 0x0000a570 ,  0x660018d4   }, // BB_tx_gain_tab_29 
+  { 0x0000a574 ,  0x660018d4   }, // BB_tx_gain_tab_30 
+  { 0x0000a578 ,  0x660018d4   }, // BB_tx_gain_tab_31 
+  { 0x0000a57c ,  0x660018d4   }, // BB_tx_gain_tab_32 
+  { 0x0000a600 ,  0x00000000   }, // BB_caltx_gain_set_0 
+  { 0x0000a604 ,  0x00000000   }, // BB_caltx_gain_set_2 
+  { 0x0000a608 ,  0x00000000   }, // BB_caltx_gain_set_4 
+  { 0x0000a60c ,  0x03804000   }, // BB_caltx_gain_set_6 
+  { 0x0000a610 ,  0x0300ca02   }, // BB_caltx_gain_set_8 
+  { 0x0000a614 ,  0x00000e04   }, // BB_caltx_gain_set_10 
+  { 0x0000a618 ,  0x03014000   }, // BB_caltx_gain_set_12 
+  { 0x0000a61c ,  0x00000000   }, // BB_caltx_gain_set_14 
+  { 0x0000a620 ,  0x00000000   }, // BB_caltx_gain_set_16 
+  { 0x0000a624 ,  0x03014000   }, // BB_caltx_gain_set_18 
+  { 0x0000a628 ,  0x03804c05   }, // BB_caltx_gain_set_20 
+  { 0x0000a62c ,  0x0701de06   }, // BB_caltx_gain_set_22 
+  { 0x0000a630 ,  0x07819c07   }, // BB_caltx_gain_set_24 
+  { 0x0000a634 ,  0x0701dc07   }, // BB_caltx_gain_set_26 
+  { 0x0000a638 ,  0x0701dc07   }, // BB_caltx_gain_set_28 
+  { 0x0000a63c ,  0x0701dc07   }, // BB_caltx_gain_set_30 
+  { 0x0000b2dc ,  0xfffd5aaa   }, // BB_cl_map_0_b1 
+  { 0x0000b2e0 ,  0xfffe9ccc   }, // BB_cl_map_1_b1 
+  { 0x0000b2e4 ,  0xffffe0f0   }, // BB_cl_map_2_b1 
+  { 0x0000b2e8 ,  0xfffcff00   }, // BB_cl_map_3_b1 
+  { 0x00016044 ,  0x010002d4   }, // ch0_TXRF2 
+  { 0x00016048 ,  0x66482400   }, // ch0_TXRF3 
+  { 0x00016280 ,  0x01000015   }, // ch0_TOP 
+  { 0x00016444 ,  0x010002d4   }, // ch1_TXRF2 
+  { 0x00016448 ,  0x66482400   }, // ch1_TXRF3 
+};
+/* END:   Added by vicharl,openwrt  2014/8/28 */
+
 #endif /* INITVALS_953X_H */
--- a/drivers/net/wireless/ath/ath9k/ar9003_eeprom.c	2015-01-26 10:48:30.504035103 +0800
+++ b/drivers/net/wireless/ath/ath9k/ar9003_eeprom.c	2015-01-26 09:40:19.588163188 +0800
@@ -3535,7 +3535,7 @@
 {
 	int bias = ar9003_modal_header(ah, is2ghz)->xpaBiasLvl;
 
-	if (AR_SREV_9485(ah) || AR_SREV_9330(ah) || AR_SREV_9340(ah))
+	if (AR_SREV_9485(ah) || AR_SREV_9330(ah) || AR_SREV_9340(ah)||AR_SREV_9531(ah))//vicharl added 9531
 		REG_RMW_FIELD(ah, AR_CH0_TOP2, AR_CH0_TOP2_XPABIASLVL, bias);
 	else if (AR_SREV_9462(ah) || AR_SREV_9550(ah) || AR_SREV_9565(ah))
 		REG_RMW_FIELD(ah, AR_CH0_TOP, AR_CH0_TOP_XPABIASLVL, bias);
--- a/drivers/net/wireless/ath/ath9k/ar9003_hw.c	2015-01-26 10:48:30.468035104 +0800
+++ b/drivers/net/wireless/ath/ath9k/ar9003_hw.c	2015-01-26 09:40:19.232163200 +0800
@@ -511,10 +511,23 @@
 	else if (AR_SREV_9550(ah))
 		INIT_INI_ARRAY(&ah->iniModesTxGain,
 			ar955x_1p0_modes_xpa_tx_gain_table);
-	else if (AR_SREV_9531(ah))
-		INIT_INI_ARRAY(&ah->iniModesTxGain,
-			qca953x_1p0_modes_xpa_tx_gain_table);
-	else if (AR_SREV_9580(ah))
+	else if (AR_SREV_9531(ah)){
+		/* BEGIN: Modified by vicharl,openwrt 2015/1/4 */
+		if(AR_SREV_9531_20(ah)){
+			   printk("tx gain table 2.0\n");
+			   INIT_INI_ARRAY(&ah->iniModesTxGain,
+				qca953xModes_w_xpa_tx_gain_table_honeybee_2p0);
+		}else if(AR_SREV_9531_11(ah)){	
+			  printk("tx gain table 1.1\n");
+			INIT_INI_ARRAY(&ah->iniModesTxGain,
+				qca953xModes_w_xpa_tx_gain_table_honeybee_1p1);// qca953x_1p0_modes_xpa_tx_gain_table
+		}else{
+			  printk("tx gain table 1.0\n");
+			INIT_INI_ARRAY(&ah->iniModesTxGain,
+           			qca953xModes_xpa_tx_gain_table_honeybee_1p0);
+		}
+		/* END:   Modified by vicharl,openwrt 2015/1/4 */
+	}else if (AR_SREV_9580(ah))
 		INIT_INI_ARRAY(&ah->iniModesTxGain,
 			ar9580_1p0_lowest_ob_db_tx_gain_table);
 	else if (AR_SREV_9462_21(ah))
@@ -595,6 +608,11 @@
 	else if (AR_SREV_9580(ah))
 		INIT_INI_ARRAY(&ah->iniModesTxGain,
 			ar9580_1p0_low_ob_db_tx_gain_table);
+	/* BEGIN: Added by vicharl,openwrt  2014/8/29 */
+	else if (AR_SREV_9531(ah)) 
+		INIT_INI_ARRAY(&ah->iniModesTxGain,
+			qca953xModes_w_xpa_tx_gain_table_honeybee_1p1);
+	/* END:   Added by vicharl,openwrt  2014/8/29 */
 	else if (AR_SREV_9565_11(ah))
 		INIT_INI_ARRAY(&ah->iniModesTxGain,
 			       ar9565_1p1_modes_low_ob_db_tx_gain_table);
--- a/drivers/net/wireless/ath/ath9k/reg.h	2015-01-26 10:48:30.536035102 +0800
+++ b/drivers/net/wireless/ath/ath9k/reg.h	2015-01-26 09:40:19.828163181 +0800
@@ -814,6 +814,9 @@
 #define AR_SREV_REVISION_9531_10        0
 #define AR_SREV_REVISION_9531_11        1
 
+/* BEGIN: Added by vicharl,openwrt  2015/1/4 */
+#define AR_SREV_REVISION_9531_20		2
+/* END:   Added by vicharl,openwrt  2015/1/4 */
 #define AR_SREV_5416(_ah) \
 	(((_ah)->hw_version.macVersion == AR_SREV_VERSION_5416_PCI) || \
 	 ((_ah)->hw_version.macVersion == AR_SREV_VERSION_5416_PCIE))
@@ -965,6 +968,11 @@
 #define AR_SREV_9531_11(_ah) \
 	(((_ah)->hw_version.macVersion == AR_SREV_VERSION_9531) && \
 	 ((_ah)->hw_version.macRev == AR_SREV_REVISION_9531_11))
+/* BEGIN: Added by vicharl,openwrt  2015/1/4 */
+#define AR_SREV_9531_20(_ah) \
+	(((_ah)->hw_version.macVersion == AR_SREV_VERSION_9531) && \
+	 ((_ah)->hw_version.macRev == AR_SREV_REVISION_9531_20))
+/* END:   Added by vicharl,openwrt  2015/1/4 */
 
 /* NOTE: When adding chips newer than Peacock, add chip check here */
 #define AR_SREV_9580_10_OR_LATER(_ah) \
